test/CheckedC/dynamic_check-code-gen.c

```diff
@@ -0,0 +1,48 @@
+// Tests for Code Generation with Checked C Extension
+// This makes sure we're generating something sensible for _Dynamic_check
+// invocations.
+//
+// RUN: %clang_cc1 -fcheckedc-extension -emit-llvm  %s -o - | FileCheck %s
+
+
+// Function with single dynamic check
+void f1(void) {
+  // CHECK: void @f1()
+  _Dynamic_check(0);
+
+  // Branch
+  // CHECK: br i1
+  // CHECK-SAME: label %_Dynamic_check_succeeded
+  // CHECK-SAME: label %_Dynamic_check_failed
+
+  // Ordering of Basic Blocks
+  // CHECK: {{^}}_Dynamic_check_succeeded
+  // CHECK: ret void
+
+  // Contents of Fail Basic Block
+  // CHECK: {{^}}_Dynamic_check_failed
+  // CHECK-NEXT: call void @llvm.trap()
+  // CHECK-NEXT: unreachable
+}
+
+// Function with two dynamic checks
+void f2(void) {
+  // CHECK: void @f2()
+
+  int i;
+  _Dynamic_check(i != 3);
+  // CHECK: icmp ne
```

Comment:
I decided to only check for the first not-equals comparison, as a way of making sure our check was happening, and not look for the second one, as it felt too much of an implementation detail.

Turns out, there is a way to make sure we don't do two comparisons, buried deep in the code generator (the API surface area is huge), I'm now using it, changes incoming.